File: upload_server.py
#!/usr/bin/env python3
"""
Servidor simples para upload de arquivos
Executar com: python upload_server.py
"""

import os
import json
import re
import uuid
from flask import Flask, request, jsonify
from flask_cors import CORS
from werkzeug.utils import secure_filename

app = Flask(__name__)
CORS(app)  # Habilitar CORS para todas as rotas

UPLOAD_FOLDER = os.path.join(os.path.dirname(__file__), 'public', 'uploads')
ALLOWED_EXTENSIONS = {'png', 'jpg', 'jpeg', 'gif', 'mp4', 'mov', 'avi', 'webp'}

# Criar pasta de uploads se não existir
os.makedirs(UPLOAD_FOLDER, exist_ok=True)

app.config['UPLOAD_FOLDER'] = UPLOAD_FOLDER
app.config['MAX_CONTENT_LENGTH'] = 50 * 1024 * 1024  # 50MB max

def allowed_file(filename):
    return '.' in filename and filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS

def slugify(text):
    """
    Converte um texto em um formato 'slug' amigável para URLs.
    Ex: "Olá Mundo!" -> "ola-mundo"
    """
    if not text:
        return ''
    # Remove acentos e caracteres especiais
    text = text.lower()
    text = re.sub(r'[àáâãäå]', 'a', text)
    text = re.sub(r'[èéêë]', 'e', text)
    text = re.sub(r'[ìíîï]', 'i', text)
    text = re.sub(r'[òóôõö]', 'o', text)
    text = re.sub(r'[ùúûü]', 'u', text)
    text = re.sub(r'[ç]', 'c', text)
    # Remove caracteres não alfanuméricos e substitui espaços por hífen
    text = re.sub(r'[^a-z0-9\s-]', '', text).strip()
    text = re.sub(r'\s+', '-', text)
    # Limita o comprimento para evitar nomes de arquivo excessivamente longos
    return text[:80]

@app.route('/upload', methods=['POST'])
def upload_file():
    if 'file' not in request.files:
        return jsonify({'error': 'Nenhum arquivo enviado'}), 400

    file = request.files['file']
    base_name = request.form.get('baseName', '')

    if file.filename == '':
        return jsonify({'error': 'Nome de arquivo vazio'}), 400

    if file and allowed_file(file.filename):
        original_filename = secure_filename(file.filename)
        ext = original_filename.rsplit('.', 1)[1].lower()
        
        if base_name:
            slug = slugify(base_name)
            short_id = uuid.uuid4().hex[:6]
            unique_filename = f"{slug}-{short_id}.{ext}"
        else:
            # Fallback para o método antigo se nenhum nome base for fornecido
            unique_filename = f"{uuid.uuid4().hex}.{ext}"

        filepath = os.path.join(app.config['UPLOAD_FOLDER'], unique_filename)
        file.save(filepath)

        # Retornar caminho relativo para o arquivo
        relative_path = f"/uploads/{unique_filename}"

        return jsonify({
            'success': True,
            'filepath': relative_path,
            'filename': unique_filename
        })

    return jsonify({'error': 'Tipo de arquivo não permitido'}), 400

@app.route('/health', methods=['GET'])
def health():
    return jsonify({'status': 'OK'})

if __name__ == '__main__':
    print("🚀 Servidor de upload iniciado na porta 5000")
    print(f"📁 Arquivos salvos em: {UPLOAD_FOLDER}")
    app.run(host='0.0.0.0', port=5000, debug=True)

File: upload-server.js
const express = require('express');
const multer = require('multer');
const sharp = require('sharp');
const path = require('path');
const crypto = require('crypto');
const fs = require('fs');

const app = express();
const PORT = 5000;

// --- Configuração ---
const UPLOAD_FOLDER = path.join(__dirname, 'public', 'uploads');
const ALLOWED_IMAGE_EXTENSIONS = new Set(['png', 'jpg', 'jpeg', 'gif', 'webp']);
const ALLOWED_VIDEO_EXTENSIONS = new Set(['mp4', 'mov', 'avi']);

// Garante que a pasta de uploads exista
fs.mkdirSync(UPLOAD_FOLDER, { recursive: true });

// --- Funções Utilitárias ---

function slugify(text) {
    if (!text) return '';
    return text
        .toString()
        .toLowerCase()
        .trim()
        .replace(/\s+/g, '-')           // Substitui espaços por -
        .replace(/[^\w\-]+/g, '')       // Remove caracteres inválidos
        .replace(/\-\-+/g, '-');        // Substitui múltiplos - por um único -
}

function getFileExtension(filename) {
    return filename.split('.').pop().toLowerCase();
}

// --- Middlewares ---

// Habilita CORS para permitir requisições do frontend
app.use((req, res, next) => {
    res.header('Access-Control-Allow-Origin', '*');
    res.header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    res.header('Access-Control-Allow-Headers', 'Content-Type, Authorization');
    next();
});

// Configuração do Multer para guardar arquivos em memória para processamento
const storage = multer.memoryStorage();
const upload = multer({ 
    storage: storage,
    limits: { fileSize: 50 * 1024 * 1024 } // Limite de 50MB
});

// --- Rotas ---

app.get('/health', (req, res) => {
    res.status(200).json({ status: 'OK' });
});

app.post('/upload', upload.single('file'), async (req, res) => {
    try {
        if (!req.file) {
            return res.status(400).json({ error: 'Nenhum arquivo enviado' });
        }

        const baseName = req.body.baseName || '';
        const originalExt = getFileExtension(req.file.originalname);
        
        const isImage = ALLOWED_IMAGE_EXTENSIONS.has(originalExt);
        const isVideo = ALLOWED_VIDEO_EXTENSIONS.has(originalExt);

        if (!isImage && !isVideo) {
            return res.status(400).json({ error: 'Tipo de arquivo não permitido' });
        }

        let uniqueFilename;
        let finalBuffer;
        let finalExtension;

        const slug = slugify(baseName).substring(0, 80);
        const shortId = crypto.randomBytes(3).toString('hex');

        if (isImage) {
            finalExtension = 'webp';
            uniqueFilename = `${slug || 'imagem'}-${shortId}.${finalExtension}`;
            
            finalBuffer = await sharp(req.file.buffer)
                .resize({ width: 1920, withoutEnlargement: true }) // Redimensiona se for maior que 1920px
                .webp({ quality: 80 }) // Converte para WebP com 80% de qualidade
                .toBuffer();

        } else { // isVideo
            finalExtension = originalExt;
            uniqueFilename = `${slug || 'video'}-${shortId}.${finalExtension}`;
            finalBuffer = req.file.buffer;
        }

        const filepath = path.join(UPLOAD_FOLDER, uniqueFilename);
        await fs.promises.writeFile(filepath, finalBuffer);

        const relativePath = `/uploads/${uniqueFilename}`;
        
        return res.status(200).json({
            success: true,
            filepath: relativePath,
            filename: uniqueFilename
        });

    } catch (error) {
        console.error('Erro no upload:', error);
        if (error instanceof multer.MulterError) {
            return res.status(413).json({ error: `Arquivo muito grande: ${error.message}` });
        }
        return res.status(500).json({ error: 'Erro interno ao processar o arquivo.' });
    }
});

// --- Iniciar Servidor ---
app.listen(PORT, '0.0.0.0', () => {
    console.log(`🚀 Servidor de upload (Node.js) iniciado na porta ${PORT}`);
    console.log(`📁 Arquivos salvos em: ${UPLOAD_FOLDER}`);
});
